examples/getting_started/web_search.ipynb

```diff
@@ -0,0 +1,255 @@
+        "\n",
+        "## Table of Contents\n",
+        "1. [Introduction](#introduction)\n",
+        "2. [Setting Up the Environment](#setup)\n",
+        "3. [Implementing the Tools](#tools)\n",
+        "4. [Building the Web Search Agent](#building)\n",
+        "5. [Running the Agent](#running)\n",
+        "6. [Conclusion](#conclusion)\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 1. Introduction <a name=\"introduction\"></a>\n",
+        "\n",
+        "The Web Search Agent is an AI-powered tool that combines the capabilities of Large Language Models (LLMs) with web search functionality. It can understand user queries, perform web searches, parse webpage content, and generate informative responses.\n",
+        "\n",
+        "Key components of our implementation include:\n",
+        "- Mirascope library for LLM interactions\n",
+        "- OpenAI's GPT model for natural language processing\n",
+        "- Custom tools for web searching and webpage parsing\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 2. Setting Up the Environment <a name=\"setup\"></a>\n",
+        "\n",
+        "First, we need to install the required packages and import the necessary modules."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "!pip install \"mirascope[openai]\" requests beautifulsoup4 pydantic"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "import os\n",
+        "\n",
+        "import requests\n",
+        "from bs4 import BeautifulSoup\n",
+        "from openai.types.chat import ChatCompletionMessageParam\n",
+        "from pydantic import BaseModel\n",
+        "\n",
+        "from mirascope.core import openai, prompt_template\n",
+        "\n",
+        "# Set your OpenAI API key\n",
+        "os.environ[\"OPENAI_API_KEY\"] = \"your-api-key-here\""
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## 3. Implementing the Tools <a name=\"tools\"></a>\n",
+        "\n",
+        "Now, let's implement our tools: `web_search` and `parse_webpage`. These functions will allow our agent to search the web and extract content from webpages."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "def web_search(query: str) -> str:\n",
+        "    \"\"\"Perform a web search and return the results.\n",
+        "\n",
+        "    This is a simplified implementation. In a real-world scenario,\n",
+        "    you would integrate with a proper search engine API.\n",
+        "    \"\"\"\n",
+        "    # Placeholder implementation\n",
```

Comment:
i don't like this. there's no reason not to show the implementation using duckduckgo with an explanation of the implementation, otherwise the final cells for running the agent won't actually work, which detracts from the value